src/main.rs

```diff
@@ -0,0 +1,278 @@
+#[macro_use]
+extern crate chan;
+extern crate chan_signal;
+use chan_signal::Signal;
+use std::thread::spawn;
+
+#[macro_use]
+extern crate clap;
+
+#[macro_use]
+extern crate log;
+extern crate env_logger;
+use log::LogLevelFilter;
+use env_logger::LogBuilder;
+
+use std::convert::From;
+
+extern crate irc;
+use irc::client::prelude::*;
+use std::default::Default;
+use std::str::FromStr;
+
+extern crate isatty;
+use isatty::stdout_isatty;
+
+use std::io::prelude::*;
+use std::io::BufReader;
+use std::io::Error as IoError;
+use std::io::stdout;
+
+
+/// Actions that threads can take
+#[derive(Debug)]
+enum Action {
+    Quit,
+    Join,
+    IoError(IoError),
+    ParseError(&'static str),
+}
+
+impl From<IoError> for Action {
+    fn from(err: IoError) -> Action {
+        Action::IoError(err)
+    }
+}
+
+impl From<&'static str> for Action {
+    fn from(err: &'static str) -> Action {
+        Action::ParseError(err)
+    }
+}
+
+
+fn main() {
+    // Catch signals we expect to exit cleanly from
+    // I don't think rust actually sends SIGPIPE, but we'll catch it anyway
+    let signal = chan_signal::notify(&[Signal::INT, Signal::TERM, Signal::PIPE]);
+
+
+    // Parse args
+    let matches = clap_app!
+        (hashpipe =>
+         (version: crate_version!())
+         (author: crate_authors!())
+         (about: "#|: Pipes data to and from an IRC connection")
+         (@arg server: -s --server +required +takes_value "IRC server to connect to")
+         (@arg port: -p --port +takes_value "Port to use (default: 6667, or 6697 with SSL)")
+         (@arg ssl: -e --ssl "Enable SSL encryption")
+         (@arg nick: -n --nick +takes_value "Nickname to use (default: hashpipe)")
+         (@arg channels: -c --channels +takes_value "Channel(s) to speak in \
+          (defalt: #hashpipe, or nothing if using raw input)")
+         // NOTE: long() is a required workaround for parsing long options with
+         // hyphens; see https://github.com/kbknapp/clap-rs/issues/321
+         (@arg raw_out: -o long("--raw-out") "Echo everything from the IRC server directly")
+         (@arg raw_in: -i long("--raw-in") "Interpret STDIN as raw IRC commands")
+         (@arg v: -v +multiple "Verbosity (1 for info, 2 for debug)")
+         (@arg quiet: -q "Only print errors (overrides -v; overridden by raw output)")
+        )
+        .get_matches();
+
+    let raw_out = matches.is_present("raw_out");
+    let raw_in = matches.is_present("raw_in");
+    let quiet = matches.is_present("quiet");
+
+    let nick = matches.value_of("nick").unwrap_or("hashpipe").to_string();
+    let server = matches.value_of("server").unwrap().to_string();
+    let ssl = matches.is_present("ssl");
+    let port = matches.value_of("port").and_then(|p| p.parse().ok());
+
+    let channels: Vec<String> = match matches.value_of("channels") {
+        Some(chans) => chans.split(",").map(|x| x.to_string()).collect(),
```

Comment:
So, cool trick, I can set `--channels ""` and 'channels' here will end up being `[""]` not `[]`, so it'll think it needs to join a channel but totally can't.

Solution here is either validating channels or at the least stripping empty values.